Update the mustache(5) manpage to the latest specification

jgonggrijp opened the pull request:
These changes aim to make the manpage an accurate, implementation-agnostic overview of the Mustache language again. In a nutshell, I made the following additions and changes:

- Mention of the specification in the DESCRIPTION section.
- Mention of dotted names, implicit iterators and lambdas in the Variables section, with a distinction between required and optional lambda behavior.
- Acknowledgement of dotted names in the Sections section.
- Example of implicit iterator in the Sections Non-Empty Lists section.
- Distinction between required and optional lambda behavior in the Sections Lamdbas section.
- Whitespace changes in several example templates and outputs to more accurately reflect what the specification dictates.
- Replaced the Ruby-inspired double callback contruction of the `wrapped` function in the Sections Lambdas section by a direct pseudomethod. The previous notation led to misunderstandings, especially when it was taken too literally when porting it to other languages such as [JavaScript](https://github.com/janl/mustache.js/#functions).
- Removed the second `render` argument of the `wrapped` lambda in order to reflect the spec.
- Added Blocks and Parents sections in order to document the inheritance candidate extension (mustache/spec#125).

The outdated content of the manpage came up in the discussion of mustache/spec#125; these are essentially parallel PRs with the common aim of refreshing the Mustache documentation. Discussion and changes in either PR could potentially influence the other PR. In particular, if mustache/spec#125 is merged first, some wording with regard to the "candidate" status of the inheritance spec may need to be revised in the current PR.

I should mention that running `rake man` proved challenging. I needed to make a couple of edits in the gemspec and the Rakefile order to make it work, and it still produced error messages although the generated html and roff files did seem to update correctly. I can share those edits if desired.

*Edit to add: given the lack of response from those authorized to merge this PR, I have decided to host my own copy of the updated manual for the time being. Jump to [this comment](https://github.com/mustache/mustache/pull/266#issuecomment-1014986964) for details.*

jgonggrijp commented:
The failed build above illustrates one of the issues I ran into when trying to run `rake man`.

jgonggrijp commented:
CC @defunkt @locks in case it makes a difference.

jgonggrijp commented:
mustache/spec#125 was merged, so I'll need to update the wording in the current PR a little bit. Marking this PR as "draft" in the meanwhile.

Danappelxx commented on man/mustache.5.ron:
```diff
@@ -85,16 +92,113 @@ Output:
+
+    * {{.}}
+
+Current context:
```
This is somewhat unrelated to the PR, but the manpage in its current revision mixes Hash and Context a lot. I understand that Hash is Ruby's version of a 'dictionary' or 'object' or 'hash map', but I think it would go a long way to do away with language specific vocabulary.

Danappelxx commented on man/mustache.5.ron:
```diff
@@ -85,16 +92,113 @@ Output:
+function or lambda, this object will be invoked with zero arguments. The
+value that is returned is then used instead of the callable object itself.
+
+An **optional** part of the specification states that if the final key in
```
I am under the impression that lambdas *as a whole* are an optional part of the spec. Happy to be corrected if that's a misunderstanding.

jgonggrijp replied:
The `~lambdas` spec, as a whole, is optional. However, the non-optional `interpolation` and `sections` specs include lambda-related semantics as well:

https://github.com/mustache/spec/blob/7b09c52a149563d53156e7abcff3dde02f7b5b9d/specs/interpolation.yml#L12-L18

https://github.com/mustache/spec/blob/7b09c52a149563d53156e7abcff3dde02f7b5b9d/specs/sections.yml#L13-L22

I have interpreted this as meaning that an implementation that doesn't include the lambdas extension, should still take the return values of object methods rather than the methods themselves.

Danappelxx replied:
That's a good catch, I stand corrected. I guess I never stumbled upon it stone the spec doesn't actually test that behavior outside of the lambda spec.

Danappelxx commented on man/mustache.5.ron:
```diff
@@ -38,6 +38,13 @@ clauses, or for loops. Instead there are only tags. Some tags are
 
+The Mustache language has a [formal specification][spec]. The current manpage
+reflects version 1.1.3 of the specification, as well as the de facto
+established [candidate extension for template inheritance][inheritance].
```
```suggestion
established [candidate extensions for template inheritance][inheritance] and lambdas.
```

jgonggrijp replied:
Well, I need to change this wording. At the time of writing, the status of the inheritance spec was different from the lambdas spec (not official vs official). I might now change it into just "including the official extensions".

Danappelxx commented on man/mustache.5.ron:
```diff
@@ -146,32 +250,54 @@ Hash:
+      <b>hub</b>
+      <b>rip</b>
 
 **Lambdas**
```
I think it would be helpful to move the Lambda's section to the bottom, along with other optional extensions (just inheritance, at the moment), in order to keep them self-contained.

jgonggrijp replied:
The manpage is currently organized by tag type rather than by spec feature. I'm not necessarily opposed to changing the structure, but I'd prefer to keep the current PR focused on just making the manpage up-to-date. Regardless, I think the current structure is user-friendly.

Danappelxx commented on man/mustache.5.ron:
```diff
@@ -286,6 +412,80 @@ Can be thought of as a single, expanded template:
+
+Blocks could be thought of as template parameters or as inline partials
+that may be passed to another template. They are part of the
+not-yet-official, optional [inheritance specification][inheritance].
```
```suggestion
optional [inheritance specification][inheritance].
```
😄

Danappelxx commented on man/mustache.5.ron:
```diff
@@ -146,32 +250,54 @@ Hash:
+text. The text passed is the literal block, unrendered. `{{tags}}` will
+not have been expanded.
+
+An **optional** part of the specification states that if the final key in
```
Comment above applies here as well (the entire Lambda spec is optional, right?).

Danappelxx reviewed (changes requested):
Overall very solid! I just have a few thoughts on organization, as well as nit-picks for wording.

jgonggrijp commented on man/mustache.5.ron:
```diff
@@ -85,16 +92,113 @@ Output:
+**Implicit Iterator**
+
+As a special case, if the `name` consists of only a dot and nothing else,
+the value that **is** the current context is interpolated as a whole. This
```
I couldn't make up my mind about which part of the sentence was most deserving of emphasis, so I decided to just remove the emphasis altogether.

jgonggrijp commented on man/mustache.5.ron:
```diff
@@ -85,16 +92,113 @@ Output:
 
+**Dotted Names**
+
+If the `name` contains dots, it is split on the dots to obtain multiple
```
I've seen it in the context of Underscore (JavaScript), but I prefer wording it a bit more elaborately to ensure that everyone understands.

jgonggrijp commented:
@Danappelxx @defunkt @locks Un-marking as draft; the manpage is now up-to-date again with the latest status of the spec, and I think I addressed all review comments.

jgonggrijp commented:
@defunkt @locks @nono @bobthecow @bruno- Can we merge this somehow?

nono commented:
Sorry, I don't have any rights on this repository

jgonggrijp commented:
@nono Thanks for responding anyway. Do you perhaps have off-GitHub means of reaching out to those who do? So far, they have been unresponsive.

suy commented on man/mustache.5.ron:
```diff
@@ -286,6 +411,80 @@ Can be thought of as a single, expanded template:
+Template:
+
+    {{<article}}
+      Never shown
```
IMHO, this could benefit from some clarification I wasn't getting why this is supposed to never show. After a lot of head-scratching, it seems like you "call" a parent, and then the content of the parent gets included in the template. You can override the blocks by re-defining them, but everything which is not a block redefinition gets ignored. That seems a bit confusing without clarification. I'm not even sure it works this way.

How about expanding the second paragraph as this:

> Like an `{{>article}}` partial, a parent lets you expand another template
> inside the current one. Unlike a partial, a parent also lets you override
> blocks of the other template. The parent can only override blocks, and
> anything else outside a block is discarded and not rendered.

jgonggrijp replied:
Good suggestion, thanks!

Everything between the `{{<parent}}{{/parent}}` tags is considered a comment except for the blocks.

jgonggrijp replied:
@suy At second thought, it seems that the next paragraph already explains this:

https://github.com/mustache/mustache/blob/d46a7c5411bcefe85e7205c3dce037cac151ef0c/man/mustache.5.ron#L451-L452

Doesn't that sufficiently address the confusion?

suy replied:
Now that you point it at it, and with the knowledge of what it does, yes. I read it "other content is ignored like if it were a comment". Before I was reading it like "other content, like comments, is ignored". Feel free to ignore my remark. I'm just a person who is not used to use templates, so maybe my experience doesn't translate to others. And it seems I got the meaning right anyway, just without much confidence. Thanks!

jgonggrijp replied:
Welcome. Out of curiosity, how does a person who is not used to using templates end up reviewing a pull request for the manpage of a templating language? :-)

suy replied:
A bit of a niche case... :) There is a [scripting language for a very old video game engine](https://www.pocketplane.net/tutorials/simscript.html) with pretty severe limitations. I did a first attempt at making a script for it with some minimal abstraction using the C/C++ preprocessor. But someone suggested using a template language, and here I am, shopping for one. I remember I've heard of mustache eons ago, and despite that the website and the Ruby implementation are not in a great shape, I've not liked the alternatives. Luckily I've found [a C++ one](https://github.com/jamboree/bustache) that seems it will be fine (supports the inheritance part that I was trying to understand).

jgonggrijp commented:
@pvande @judofyr From the network graph, it appears that both of you have merged PRs in this repo in the distant past. Is either of you still in a position to do this?

judofyr commented:
I don't have any permissions either unfortunately :-(

jgonggrijp commented:
Oh well. Thanks for the heads-up, @judofyr.

Most likely I'll end up hosting my own copy of the updated version, at least until there is movement on this repo again. I'll report back here when I've made the move.

jgonggrijp commented:
@pietroppeter @suy @HamidrezaSafari @Danappelxx @defunkt @locks @nono @judofyr From now on I'm hosting an up-to-date version of `man 5 mustache` over here: https://jgonggrijp.gitlab.io/wontache/mustache.5.html. Feel free to share that link if you want to give people an up-to-date reference of the templating language. I accept pull requests to my fork if needed, although I'm also likely to proactively update that page myself.

To clarify, that page is part of the website of my own Mustache implementation for JavaScript, which is called "Wontache". Wontache is still very new, I haven't even published it to NPM (the package index for JS) yet.

pietroppeter commented:
Just noticed this comment about you hosting, well done! FYI I was interested in this thread because I had "translated" that man page to be used with a Nim mustache implementation: https://pietroppeter.github.io/nimib/mostaccio.html

I at some point I will likely update my document to use your version, thanks for having worked on this.

jgonggrijp commented:
Nice to meet @pietroppeter! Also, welcome!

jgonggrijp commented:
@locks Thanks for merging!

jgonggrijp commented:
Update to the official website being outdated: it is much less outdated now! [The mustache(5) manpage](https://mustache.github.io/mustache.5.html) is now current with version 1.3.0 of the spec. It does not mention dynamic names yet, but I can now merge updates to the website myself thanks to @locks and I will address the remaining bits of outdatedness sometime in the near future.
